### ISSUE TITLE
Loops in legacy GLES

### ISSUE THREAD

**Issue by luboslenco** [KhronosGroup/SPIRV-Cross#55]
For loops are currently broken in legacy GLES, as loop variable must be declared inside loop scope. Throwing "for: Missing init declaration":

``` glsl
i = 0;
for (; (i < 4); i = (i + 1))
{
    amount = (amount * 0.5);
}
```

Any insight on how hard would this be to fix?

**Comment by HansKristian-ARM** [comment]
Ouch, looking at the ESSL 1.0 spec it is extremely strict about for loops, not even i = i + 1 will work.
I estimate fixing this would be quite difficult, even being able to handle for loops this way in SPIRV-Cross was fairly hard.

One way to solve the i = 0 issue is to analyze in which scope variables are actually accessed. In this case, it might be possible to move the int declaration inside the for loop. Fixing this would also make for cleaner variable declaration as variable declarations can currently only happen in the entry block.

i = 0 is done in the block before the loop header, so we must also forward the i = 0 assignment to the first block which actually uses it, which also gets pretty annoying. Analyzing which variables are accessed in a for loop header and treat them specially could work, but that's kinda ugly as well.

Fixing i = i + 1 to i++ is fairly painful as a series of SPIR-V opcodes need to be analyzed and a peep-hole optimization must be made to reduce i = i + 1 to i++ or i += 1, which is what ESSL 1.0 seems to support.
Since the increment needs to happen inside a continue block, it should be possible to reduce the scope for which opcodes to look for and which cases to consider.

One possible solution with say OpIAdd is to propagate the backing variable into the SPIRExpression, and on a store op, evaluate if we're writing back to the variable. If so, use the += form instead of i + 1, still quite ugly though. Then there is OpPhi ...

In short, this is pretty hard, but finding a good workaround seems important.

**Comment by luboslenco** [comment]
Ah, feared of this. Looks like an ugly task, even more so as it can pollute existing codebase. Here is hoping for WebGL2 spreading.

**Comment by HansKristian-ARM** [comment]
Which ESSL compiler are you having issues with, and does i = i + 1 at least work?

**Comment by luboslenco** [comment]
I am using SPIRV-Cross through [Krafix](https://github.com/KTXSoftware/krafix) (https://github.com/KhronosGroup/glslang), and running in Chromium. 

Unfortunately not even i = i + 1 does work. Quick test at [shdr](http://shdr.bkcore.com/#1/dVFNb8IwDP0rVk7tYOVj7ELHadJum3baZUwotKEYtUmVpIwP8d/npC0wRA+RFb9n+/n5yFKVVIWQ1rDpNyu1SNCgkrDGbF3CKlfcxnNZSVwpXdR/sFiIq+RWJGPQwqi8slRKyJbrPcrMIU+w+lQG7wIfVM1zShOgMIWCowzCuTzOJYB7zThhLMxgFA1jn1QaApQWkJLDmMILPFPo9ULwlYNBF4V+CD0YNUSoWz8SJxr73qd2cpYv3jTPXlVOrWZO7yRo9MKDL+s7QSFVnVi/0zdurcZlZUW9cXlx4gaRrRWtqQW3bfadE3X3H5tAoVKRf6H4vQ+XWm1E4oad8Q7zO4917yqtB7NGGx5EcK2SzKm/offT+ea2Jv6NXiJ6U1tLznbCRQNVERzt9oe4uck1crMgNSSyv8dPn/lpbFpHMzDVgUspFu460caw0x8=).

**Comment by HansKristian-ARM** [comment]
To update on this issue, I've started looking into this after some thinking on and off. The first step is to implement a simple CFG to be able to query which scope variables live in.

**Comment by HansKristian-ARM** [comment]
Github closed this automatically, not quite there yet.

**Comment by HansKristian-ARM** [comment]
Got an initial implementation of this: https://github.com/KhronosGroup/SPIRV-Cross/tree/for-loop-fix.
It analyzes the CFG to find loop variables, and can promote them to a for loop initializer if possible. Still need to optimize i = i + 1 -> i += 1 to make it valid GLES2, but that seems like a fairly simple task now compared to this ...
